ocaml-lsp-server/src/dune.ml

```diff
@@ -1,46 +1,158 @@
+    end)
+    (Chan)
+
 type state =
   | Binary_not_found
   | Out_of_date
-  | Running
 
 type t =
   { scheduler : Scheduler.t
-  ; state : state Fiber.Ivar.t
+  ; diagnostics : Diagnostics.t
   }
 
-let create scheduler = { scheduler; state = Fiber.Ivar.create () }
+let create scheduler diagnostics = { scheduler; diagnostics }
 
-let state t : state Fiber.t =
-  let open Fiber.O in
-  let* state = Fiber.Ivar.peek t.state in
-  match state with
-  | Some s -> Fiber.return s
-  | None -> (
-    match Bin.which "dune" with
-    | None ->
-      let res = Binary_not_found in
-      let+ () = Fiber.Ivar.fill t.state res in
-      res
-    | Some bin ->
-      let bin = Fpath.to_string bin in
-      let dev_null () = Unix.openfile "/dev/null" [ Unix.O_RDWR ] 0o666 in
-      let stdin = dev_null () in
-      let stdout = dev_null () in
-      let stderr = dev_null () in
+let lsp_of_dune dune =
+  let module D = Drpc.Diagnostic in
+  let range =
+    match D.loc dune with
+    | None -> Range.first_line
+    | Some loc ->
+      let loc =
+        { Loc.loc_start = loc.start; loc_end = loc.stop; loc_ghost = false }
+      in
+      Range.of_loc loc
+  in
+  let severity =
+    D.severity dune
+    |> Option.map ~f:(function
+         | D.Error -> DiagnosticSeverity.Error
+         | Warning -> DiagnosticSeverity.Warning)
+  in
+  let message = Format.asprintf "%a@." Pp.to_fmt (D.message dune) in
+  Diagnostic.create ~range ?severity ~source:"dune" ~message ()
 
+let run_rpc t bin =
+  let stdin_i, stdin_o = Unix.pipe () in
+  let stdout_i, stdout_o = Unix.pipe () in
+  let pid =
+    let args = Array.of_list [ bin; "rpc"; "init" ] in
+    (* TODO collect stderr and report it properly *)
+    Unix.create_process bin args stdin_i stdout_o Unix.stderr
+    |> Stdune.Pid.of_int
+  in
+  let chan = Chan.create t.scheduler stdout_i stdin_o in
+  Fiber.fork_and_join_unit
+    (fun () ->
       let open Fiber.O in
-      let pid =
-        let args = Array.of_list [ bin; "rpc"; "--help=plain" ] in
-        Unix.create_process bin args stdin stdout stderr |> Stdune.Pid.of_int
-      in
-      let* status = Scheduler.wait_for_process t.scheduler pid in
-      let state =
-        match status with
-        (* TODO actually turn on *)
-        | Unix.WEXITED 0 -> Running
-        | _ -> Out_of_date
-      in
-      let+ () = Fiber.Ivar.fill t.state state in
-      state)
+      let* res = Scheduler.wait_for_process t.scheduler pid in
+      match res with
+      | Unix.WEXITED 0 -> Fiber.return ()
+      | _ ->
+        (* TODO inform client instead *)
+        failwith "Unexpected termination from dune")
+    (fun () ->
+      Client.connect_persistent chan
+        ~on_connected:(fun () t ->
+          let sub = Drpc.Subscribe.Diagnostics in
+          Client.notification t Drpc.Notification.subscribe sub)
+        ~on_connect:(fun () ->
+          let init =
+            Drpc.Initialize.create ~id:(Drpc.Id.make (Atom "ocamllsp"))
```

Comment:
Should the ID of the session be dynamic? In case the connection is lost and we have to re-create one, I would expect to have a new session ID.

Reply:
The session id doesn't have to be unique, it's just to have something "nice" to show in `$ dune rpc status`.